### PR TITLE
Prevent abstract casts handler for Std.string in inlined code

### LINKED ISSUE #9355: Regression: enough inlining thwarts ECheckType
Consider the following marvel:

```haxe
class Main {
  static inline function output(s:String)
    trace(s);

  static inline function writeFloat(f:Float)
    output(Std.string((f:Float)));

  static inline function writeFloatDirectly(f:Float)
    trace(Std.string((f:Float)));

  static function main() {
    var o:Opacity = .5;
    writeFloat(o);
    writeFloatDirectly(o);
    output(Std.string((o:Float)));
  }
}

abstract Opacity(Float) from Float to Float {
  @:to public function toString():String
    return 'huh?';
}
```

In 4.0.5 this outputs:

```
Main.hx:3: 0.5
Main.hx:9: 0.5
Main.hx:3: 0.5
```

In 031345d (latest available nightly at the time of writing this):

```
Main.hx:3: huh?
Main.hx:9: 0.5
Main.hx:3: 0.5
```

### PR DESCRIPTION
**Pull request by RealyUniqueName**
Fixes #9355

### CODE REVIEW

**Comment by Simn** [comment]
This sucks... :(

**Comment by RealyUniqueName** [comment]
Sucks, but what else can we do?

**Comment by RealyUniqueName** [comment]
@Simn so, what do you think about merging this?

**Comment by Simn** [comment]
There's no way this is the right fix.

I also disagree with your assessment here:

> The writeFloat(o) call gets inlined as var f:Float = o; trace(Std.string(f)); and reduced to trace(Std.string(o)), which is correct.

This reduction is not sound if it causes us to lose the intermediate type `Float`. Where does this optimization occur?

**Comment by RealyUniqueName** [comment]
Here:
https://github.com/HaxeFoundation/haxe/blob/a1a852989dff91e51ea760a21a924f8a04dcef31/src/optimization/inline.ml#L492-L499

**Comment by RealyUniqueName** [comment]
To my understanding the optimization is (almost) correct. It just should additionally change `etype` to `Float`.
I tried to do that, but it broke quite a bunch of tests.

**Comment by Simn** [comment]
Don't we lose the original type if we do that?

The problem with casts is that they aren't transitive: Casting A to B and then B to C is not the same as casting A to C. This means that we have to retain all assignments that change the type.

File: src/optimization/inline.ml [hunk 1]
```diff
@@ -513,8 +513,8 @@ class inline_state ctx ethis params cf f p = object(self)
 			| TCall({eexpr = TField(_,FStatic({cl_path=[],"Std"},{cf_name = "string"}))} as e1,[e2]) ->
 				let e2' = inline_params true false e2 in
 				let e2' =
-					if fast_eq e2.etype e2'.etype then e2'
-					else {e2' with etype = e2.etype}
+					if fast_eq (follow e2.etype) (follow e2'.etype) then e2'
+					else {e2 with eexpr = TCast (e2',None) }
```
Comment (nadako) [inline src/optimization/inline.ml:517]: shouldn't this still have the etype?

Comment (RealyUniqueName) [reply src/optimization/inline.ml:517]: `etype` here is still "taken" from `e2`.
The point of this PR is to keep the type of the argument in `Std.string` calls.

Comment (nadako) [reply src/optimization/inline.ml:517]: ah, you're right :) i thought it's `e2' with`

**Pull request merged**